File: lab-2/Model/CardModule/CardSet.js
/**
 * Randomly shuffles the order of an array.
 * 
 * @param {Array} array - an array of objects to be randomized
 */
function shuffleArray(array){
    /* 
        Source for random number algorithm: 
        https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
    */
    for(let i = 0; i < array.length * array.length; i++){
        const randIndex = Math.floor(Math.random() * array.length);
        if(array[randIndex] == undefined){
            
        }
        // removes an item at a random index an pushes it to the beginning of the array
        array.push(array.splice(randIndex, 1)[0]);
    }
}

/**
 * A class that represents a deck of cards in the game of Set.
 * 
 * @member {Array(Card)} deck - array of current cards in the deck
 * 
 * @author Adam Lechliter
 */
class CardSet {

    /**
     * Initialize a deck of 81 cards the deck (applies a normal shuffle)
     * 
     * @return {CardSet} a new instance of type CardSet
     */
    constructor(){
        this.initializeNormalDeck();
    }

    /**
     * Deals a card from the deck, removing that card from the deck. Returns null
     * if the deck is empty. 
     * 
     * @returns {Card | null} a card removed from the deck
     */
    dealCard(){
        let card = null;
        if (this.deck.length > 0){
            card = this.deck.pop();
        }
        return card;
    }

    /**
     * Initializes the deck with 81 cards and then shuffles every card in the deck
     * 
     * @returns void
     */
    initializeNormalDeck(){
        this.deck = [];
        const tempDeck = [];
        Card.getNumShapeTypes().forEach(function(numShapes){
            Card.getShapeTypes().forEach(function(shape){
                Card.getShadingTypes().forEach(function(shading){
                    Card.getColorTypes().forEach(function(color){
                        tempDeck.push(new Card(numShapes, shape, shading, color));
                    })
                })
            })
        });
        this.deck = tempDeck;
        shuffleArray(this.deck);
    }

    /**
     * Initializes the deck with 81 cards, applying the rookie 1 shuffle
     * 
     * Rookie 1 shuffle: shuffle every shading type and color
     * 
     * @returns void
     */
    initializeRookieDeck1(){
        this.deck = [];
        const tempDeck = [];

        Card.getNumShapeTypes().forEach(function(numShapes){
            Card.getShapeTypes().forEach(function(shape){
                Card.getShadingTypes().forEach(function(shading){
                    Card.getColorTypes().forEach(function(color){
                        tempDeck.push(new Card(numShapes, shape, shading, color));
                    })
                })
                shuffleArray(tempDeck);
                this.deck = this.deck.concat(tempDeck);
                tempDeck = [];
            })
        });
        this.deck = tempDeck;
    }

    /**
     * Initializes the deck with 81 cards, applying the rookie 1 shuffle
     * 
     * Rookie 2 shuffle: shuffle every shading type, color, and shape
     * 
     * @returns void
     */
    initializeRookieDeck2(){
        this.deck = [];
        const tempDeck = [];
        Card.getNumShapeTypes().forEach(function(numShapes){
            Card.getShapeTypes().forEach(function(shape){
                Card.getShadingTypes().forEach(function(shading){
                    Card.getColorTypes().forEach(function(color){
                        tempDeck.push(new Card(numShapes, shape, shading, color));
                    })
                })
            })
            shuffleArray(tempDeck);
            this.deck = this.deck.concat(tempDeck);
            tempDeck = [];
        });
        this.deck = tempDeck;
    }

    /**
     * Reports the remaining amount of cards in the deck
     * 
     * @returns {number} the number of remaing cards in the deck
     */
    getRemainingAmount(){
        return this.deck.length;
    }

    /**
     * Reports whether the deck is empty or not
     * 
     * @returns {boolean} true if the deck is empty
     */
    isEmpty(){
        return this.deck.length == 0;
    }

}

File: lab-1/CardImages/CardModule/card-set.rb
require_relative './card.rb'

module CardModule

	# a class that represents a deck of cards in the game of Set.
	#
	# @author Adam Lechliter, Jing Wen
	class CardSet

		# Initialize a deck of 81 cards and shuffles the deck
		#
		# @return [CardSet] a new instance of type CardSet
		def initialize
			@deck = []

			Card.NUM_SHAPES.each do |num|
				Card.SHAPES.each do |shape|
					Card.SHADING.each do |shading|
						Card.COLOR.each do |color|
							@deck.push Card.new(num, shape, shading, color)
						end
					end 
				end
			end
			@deck.shuffle!
		end


		def print_deck()
			@deck.each {|card|
				puts "#{card.num_shapes},#{card.shape},#{card.shading},#{card.color}"
			}
		end




		
		# public methods ----------------------------------------------------


		# Deals a card from the deck, removing that card from the deck. Returns nil
		# if the deck is empty. 
		#
		# @return [Card] a card removed from the deck
		def deal_card!
			card = nil
			if @deck.length > 0
				card = @deck.pop
			end
			card
		end

		# reshuffle the cards in normal mode
		def normal!()
			Card.NUM_SHAPES.each do |num|
				Card.SHAPES.each do |shape|
					Card.SHADING.each do |shading|
						Card.COLOR.each do |color|
							@deck.push Card.new(num, shape, shading, color)
						end
					end
				end
			end
			@deck.shuffle!
		end

		# reshuffle the cards in the rookie mode 1
		def rookie_shuffle1!()
			@deck=[]
			temp=[]
			Card.NUM_SHAPES.each do |num|
				Card.SHAPES.each do |shape|
					Card.SHADING.each do |shading|
						Card.COLOR.each do |color|
							temp.push Card.new(num, shape, shading, color)
						end
					end
					temp.shuffle!
					@deck+=temp
					temp=[]
				end
			end
		end

		# reshuffle the cards in rookie mode 2
		def rookie_shuffle2!()
			@deck=[]
			temp=[]
			Card.NUM_SHAPES.each do |num|
				Card.SHAPES.each do |shape|
					Card.SHADING.each do |shading|
						Card.COLOR.each do |color|
							temp.push Card.new(num, shape, shading, color)
						end
					end
				end
				temp.shuffle!
				@deck+=temp
				temp=[]
			end
		end


		# Reports the amount of cards left in the deck 
		#
		# @return [Number] number of cards left in the deck
		def remaining_amount
			@deck.length
		end


		# Returns whether the deck is empty or not 
		#
		# @return [Boolean] true if the deck is empty
		def is_empty?
			@deck.length == 0
		end

	end
end